File: barber-flow-api/Barber.Flow.Api/Barber.Flow.Domain/Entities/DailyReport.cs
namespace Barber.Flow.Domain.Entities;

public class DailyReport
{
    public required DateOnly ReportDate { get; init; }

    public required int TotalCustomersServed { get; init; }

    public required decimal GrossRevenue { get; init; }

    public required decimal NetProfit { get; init; }

    public required decimal CommissionAmount { get; init; }

    public required decimal FixedDailyExpense { get; init; }

    public required IReadOnlyCollection<PaymentMethodBreakdown> PaymentMethodBreakdown { get; init; }

    public required IReadOnlyCollection<CompletedAppointmentReportItem> CompletedAppointments { get; init; }

    public required DateTime GeneratedAt { get; init; }
}

public class PaymentMethodBreakdown
{
    public required string PaymentMethod { get; init; }

    public required string Label { get; init; }

    public required decimal Total { get; init; }

    public required int AppointmentCount { get; init; }
}

public class CompletedAppointmentReportItem
{
    public required string Id { get; init; }

    public required string ClientName { get; init; }

    public string? ServiceName { get; init; }

    public required string Time { get; init; }

    public required decimal ServicePrice { get; init; }

    public required string PaymentMethodUsed { get; init; }
}

File: barber-flow-mobile/Barber.Flow.Mobile/src/features/reports/dailyReport.ts
import type { Appointment, AppointmentPaymentMethod } from "../appointments/appointments.types";
import {
  APPOINTMENT_PAYMENT_METHOD_OPTIONS,
} from "../appointments/appointments.types";
import type { ReportCalculationSettings } from "../../types/settings";

export interface DailyReportPaymentBreakdown {
  appointmentCount: number;
  paymentMethod: AppointmentPaymentMethod;
  total: number;
}

export interface DailyReportSummary {
  commissionAmount: number;
  completedAppointments: Appointment[];
  fixedDailyExpense: number;
  grossRevenue: number;
  netProfit: number;
  paymentBreakdown: DailyReportPaymentBreakdown[];
  selectedDate: string;
  totalCustomersServed: number;
}

export const calculateDailyReportSummary = (
  appointments: Appointment[],
  selectedDate: string,
  settings: ReportCalculationSettings,
): DailyReportSummary => {
  const completedAppointments = appointments
    .filter(
      (appointment) =>
        appointment.date === selectedDate && appointment.status === "completed",
    )
    .sort((left, right) => left.time.localeCompare(right.time));

  const grossRevenue = completedAppointments.reduce(
    (total, appointment) => total + (appointment.servicePrice ?? 0),
    0,
  );
  const commissionAmount = grossRevenue * (settings.commissionPercentage / 100);
  const paymentBreakdown = APPOINTMENT_PAYMENT_METHOD_OPTIONS.map((paymentMethod) => {
    const paymentMethodAppointments = completedAppointments.filter(
      (appointment) => appointment.paymentMethodUsed === paymentMethod,
    );

    return {
      appointmentCount: paymentMethodAppointments.length,
      paymentMethod,
      total: paymentMethodAppointments.reduce(
        (total, appointment) => total + (appointment.servicePrice ?? 0),
        0,
      ),
    };
  });

  return {
    commissionAmount,
    completedAppointments,
    fixedDailyExpense: settings.fixedDailyExpense,
    grossRevenue,
    netProfit: grossRevenue - commissionAmount - settings.fixedDailyExpense,
    paymentBreakdown,
    selectedDate,
    totalCustomersServed: completedAppointments.length,
  };
};